### PR TITLE
Feature/smart cities lulc

### PR DESCRIPTION
**Pull request by weiqi-tori**
## Did you read the Contributor Guide?

- Yes, I have read [How to contribute](https://github.com/wri/cities-cif#How-to-contribute)

- No, I haven't read it.

## Is this PR related to a JIRA ticket?

- Yes, the URL of the associated JIRA ticket is https://gfw.atlassian.net/browse/CIF-XXX. The PR name follows the format `[CIF-XXX] my subject`.

- No, this is a documentation update. The PR name follows the format `[DOCS] my subject`.

## Did this PR include necessary documentation updates?

- Yes, I am adding a new API feature.
- Yes, I have updated the documentation update.
- No, this PR does not affect any public API so no need to change the docs.

### CODE REVIEW

**Comment by S-AI-F** [comment]
To Test: 
- layer:  OpenStreetMap, NaturalAreas
- indicators: urban_open_space(zones), natural_areas(zones)

File: city_metrix/layers/smart_cities_lulc.py [hunk 1]
```diff
@@ -11,11 +23,194 @@ def __init__(self, land_cover_class=None, **kwargs):
         self.land_cover_class = land_cover_class
 
     def get_data(self, bbox):
-       raise NotImplementedError()
+        crs = get_utm_zone_epsg(bbox)
 
+        # ESA reclass and upsample
+        esa_world_cover = EsaWorldCover().get_data(bbox)
 
+        reclass_map = {
```
Comment (jterry64) [inline city_metrix/layers/smart_cities_lulc.py:31]: I think OSM Open Space layer you moved here is also used for a different indicator, so I think you should just use that layer here instead of deleting it

File: city_metrix/layers/building_classifier.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        buildings_sample = gpd.read_file(self.geo_file)
+        buildings_sample.to_crs(epsg=4326,inplace=True)
+        bbox = buildings_sample.reset_index().total_bounds
+
+        crs = get_utm_zone_epsg(bbox)
+
+        # ESA reclass and upsample
+        esa_world_cover = EsaWorldCover().get_data(bbox)
+
+        reclass_map = {
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:32]: you define this in a couple places, could you define it all in one layer?

File: city_metrix/layers/building_classifier.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        
+        clf = DecisionTreeClassifier(max_depth=4)
+        # encode labels
+        buildings_sample['Slope_encoded'] = buildings_sample['Slope'].map({'low': 41, 'high': 42})
+        # drop records with NA in Slope
+        buildings_sample = buildings_sample.dropna(subset=['Slope'])
+
+        clf.fit(buildings_sample[['ULU', 'ANBH', 'Area_m']], buildings_sample['Slope_encoded'])
+
+        # plt.figure(figsize=(20, 10))
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:172]: remove commented out code

File: city_metrix/layers/building_classifier.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            dst_crs=crs,
+            shape=esa_1m.shape,
+            resampling=Resampling.nearest,
+            nodata=1
+        )
+
+        # ANBH is the average height of the built surfaces, USE THIS
+        # AGBH is the amount of built cubic meters per surface unit in the cell
+        # https://ghsl.jrc.ec.europa.eu/ghs_buH2023.php
+        anbh = (ee.ImageCollection("projects/wri-datalab/GHSL/GHS-BUILT-H-ANBH_R2023A")
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:85]: maybe move to its own layer and call here to simplify this class?

File: city_metrix/layers/building_classifier.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+            nodata=0
+        )
+
+        return buildings_sample, esa_1m, ulu_lulc_1m, anbh_1m, crs
+    
+
+    # Extract values to buildings as coverage fractions
+    # Extract average of pixel values to buildings
+    # Reproject to local state plane and calculate area
+    def calc_majority_ULU_mean_ANBH_area(self, row, building_sample_1m, id_col, ulu_lulc_1m, anbh_1m):
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:114]: This should be all lowercase. Also kind of hard to read with two acronyms, is there a simpler name you could use? I'm not really sure what this function is trying to get

File: city_metrix/layers/building_classifier.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        anbh_data = anbh_data.squeeze("time").transpose("Y", "X").rename({'X': 'x', 'Y': 'y'})
+        
+        anbh_1m = anbh_data.rio.reproject(
+            dst_crs=crs,
+            shape=esa_1m.shape,
+            resampling=Resampling.nearest,
+            nodata=0
+        )
+
+        return buildings_sample, esa_1m, ulu_lulc_1m, anbh_1m, crs
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:108]: This function shouldn't return 4 arrays, it should return 1 array with the result of the building classifier (that's an array with classifications, right?). Otherwise, there's no way for people to actually use this class.

File: city_metrix/layers/building_classifier.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+            vector_data=gdf,
+            measurements=["Value"],
+            like=snap_to,
+            fill=0
+        ).Value
+
+        return raster.rio.reproject_match(snap_to)
+    
+    def building_class_tree(self):
+        buildings_sample, esa_1m, ulu_lulc_1m, anbh_1m, crs = self.get_data()
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:159]: could you provide some comments on what this building classifier is classifying, and what the output is?

File: city_metrix/layers/building_classifier.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+from .urban_land_use import UrbanLandUse
+
+
+class BuildingClassifier(Layer):
+    def __init__(self, geo_file=None, **kwargs):
+        super().__init__(**kwargs)
+        self.geo_file = geo_file
+
+    def get_data(self):
+        buildings_sample = gpd.read_file(self.geo_file)
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:23]: what is this file?

File: city_metrix/layers/building_classifier.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        
+        # Extract values to buildings as coverage fractions
+        # when there is no majority class, use 1-Non-residential as default
+        if masked_ulu.size == 0:
+            majority_ULU = 1
+        else:
+            unique, counts = np.unique(masked_ulu, return_counts=True)
+            sorted_indices = np.argsort(-counts)  # Sort by descending order
+            
+            # Apply your specific logic
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:126]: what does this mean?

File: city_metrix/layers/building_classifier.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    # Extract average of pixel values to buildings
+    # Reproject to local state plane and calculate area
+    def calc_majority_ULU_mean_ANBH_area(self, row, building_sample_1m, id_col, ulu_lulc_1m, anbh_1m):
+        mask = building_sample_1m == row[id_col]
+        masked_ulu = ulu_lulc_1m.values[mask]
+        
+        # Extract values to buildings as coverage fractions
+        # when there is no majority class, use 1-Non-residential as default
+        if masked_ulu.size == 0:
+            majority_ULU = 1
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:121]: use lowercase for variables and function names

File: city_metrix/layers/building_classifier.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+
+        # Extract average of pixel values to buildings
+        masked_anbh = anbh_1m.values[mask]
+        if masked_anbh.size == 0:
+            mean_ANBH = 0
+        else:
+            mean_ANBH = np.mean(masked_anbh)
+        
+        # Reproject to local state plane and calculate area
+        Area_m = row.geometry.area
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:142]: lowercase

File: city_metrix/layers/building_classifier.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+                majority_ULU = unique[sorted_indices[0]]
+            elif len(sorted_indices) > 1:
+                majority_ULU = unique[sorted_indices[1]]
+            else:
+                majority_ULU = 1  # Default to 1 non-residential
+
+        # Extract average of pixel values to buildings
+        masked_anbh = anbh_1m.values[mask]
+        if masked_anbh.size == 0:
+            mean_ANBH = 0
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:137]: lowercase

File: city_metrix/layers/building_classifier.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    
+
+    # Extract values to buildings as coverage fractions
+    # Extract average of pixel values to buildings
+    # Reproject to local state plane and calculate area
+    def calc_majority_ULU_mean_ANBH_area(self, row, building_sample_1m, id_col, ulu_lulc_1m, anbh_1m):
+        mask = building_sample_1m == row[id_col]
+        masked_ulu = ulu_lulc_1m.values[mask]
+        
+        # Extract values to buildings as coverage fractions
```
Comment (jterry64) [inline city_metrix/layers/building_classifier.py:118]: what are the coverage fractions? the coverage of each pixel by a building? I'm a little confused on how the code below is accomplishing this. It might be better to use less complex indexing, it makes the code hard to follow.

File: city_metrix/layers/smart_cities_lulc.py [hunk 14]
```diff
@@ -11,11 +23,194 @@ def __init__(self, land_cover_class=None, **kwargs):
...
+
+
+        # Water
+        water_osm = OpenStreetMap(osm_tag=water_tag).get_data(bbox).to_crs(crs).reset_index()
+        water_osm['Value'] = 20
+        water_1m = rasterize_osm(water_osm, esa_1m)
+
+
+        # Roads
+        roads_osm = OpenStreetMap(osm_tag=roads_tag).get_data(bbox).to_crs(crs).reset_index()
```
Comment (jterry64) [inline city_metrix/layers/smart_cities_lulc.py:96]: for these last 3, perhaps make it 3 subclasses of OSM that each just pass these tags? I think there are other indicators we might to re-use these for, and would simplify this very long function

File: city_metrix/layers/smart_cities_lulc.py [hunk 15]
```diff
@@ -11,11 +23,194 @@ def __init__(self, land_cover_class=None, **kwargs):
...
+            dst_crs=crs,
+            shape=esa_1m.shape,
+            resampling=Resampling.nearest,
+            nodata=1
+        )
+
+        # ANBH is the average height of the built surfaces, USE THIS
+        # AGBH is the amount of built cubic meters per surface unit in the cell
+        # https://ghsl.jrc.ec.europa.eu/ghs_buH2023.php
+        anbh = (ee.ImageCollection("projects/wri-datalab/GHSL/GHS-BUILT-H-ANBH_R2023A")
```
Comment (jterry64) [inline city_metrix/layers/smart_cities_lulc.py:154]: as mentioned in building_classifier.py, you should just make this a separate layer, and then you wouldn't need to define this twice

File: city_metrix/layers/smart_cities_lulc.py [hunk 16]
```diff
@@ -11,11 +23,194 @@ def __init__(self, land_cover_class=None, **kwargs):
...
+            shape=esa_1m.shape,
+            resampling=Resampling.nearest,
+            nodata=0
+        )
+
+        building_osm = OpenStreetMap(osm_tag=building_tag).get_data(bbox).to_crs(crs).reset_index()
+        building_osm['Value'] = building_osm['osmid']
+        building_osm_1m = rasterize_osm(building_osm, esa_1m)
+        
+        building_osm[['ULU', 'ANBH', 'Area_m']] = building_osm.apply(lambda row: BuildingClassifier().calc_majority_ULU_mean_ANBH_area(row, building_osm_1m, 'osmid', ulu_lulc_1m, anbh_1m), axis=1)
```
Comment (jterry64) [inline city_metrix/layers/smart_cities_lulc.py:181]: shouldn't this all be happening in BuildingClassidier?

File: city_metrix/metrics.py [hunk 17]
```diff
@@ -72,7 +70,9 @@ def built_land_with_high_land_surface_temperature(zones):
 
 def urban_open_space(zones):
     built_up_land = EsaWorldCover(land_cover_class=EsaWorldCoverClass.BUILT_UP)
-    open_space = OSMOpenSpace()
+    open_space_tag = {'leisure': ['park', 'nature_reserve', 'common', 'playground', 'pitch', 'track'],
```
Comment (jterry64) [inline city_metrix/metrics.py:73]: same as above, move this to a subclass of OSM

**Review by jterry64 — CHANGES_REQUESTED** [review]

File: city_metrix/layers/building_classifier.py (hunk 4 above)
Comment (S-AI-F) [reply city_metrix/layers/building_classifier.py:85]: We will need this layer for the next month so it would be great to have it as a separate class as suggested by Justin

File: city_metrix/layers/smart_cities_lulc.py (hunk 1 above)
Comment (weiqi-tori) [reply city_metrix/layers/smart_cities_lulc.py:32]: Updated OSM layer to have predefined land use types to be reused

File: city_metrix/layers/building_classifier.py (hunk 4 above)
Comment (weiqi-tori) [reply city_metrix/layers/building_classifier.py]: Created a layer for ANBH

File: city_metrix/layers/smart_cities_lulc.py (hunk 14 above)
Comment (weiqi-tori) [reply city_metrix/layers/smart_cities_lulc.py]: Updated OSM layer to have predefined land use types to be reused

File: city_metrix/layers/smart_cities_lulc.py (hunk 15 above)
Comment (weiqi-tori) [reply city_metrix/layers/smart_cities_lulc.py]: Created a layer for ANBH

File: city_metrix/metrics.py (hunk 17 above)
Comment (weiqi-tori) [reply city_metrix/metrics.py]: Updated OSM layer to have predefined land use types to be reused

File: city_metrix/layers/building_classifier.py (hunk 8 above)
Comment (weiqi-tori) [reply city_metrix/layers/building_classifier.py:24]: A training dataset of 200 buildings https://github.com/wri/cities-heat/blob/main/data/buildings-sample-classed_LA.geojson

File: city_metrix/layers/building_classifier/building_classifier.py [hunk 18]
```diff
@@ -0,0 +1,197 @@
+import xarray as xr
```
Comment (jterry64) [inline city_metrix/layers/building_classifier/building_classifier.py:1]: is this file just a duplicate of the .py file?

**Review by jterry64 — CHANGES_REQUESTED** [review]
Looks good for the most part, just had questions about those two files

File: city_metrix/layers/building_classifier/V2-building-class-data.geojson [hunk 19]
```diff
@@ -0,0 +1,303 @@
+{
```
[the remark this answers is not in the document]
Comment (chrowe) [reply city_metrix/layers/building_classifier/V2-building-class-data.geojson:1]: It is only 184 KB so not very large. I like having it here since it is the basis for the model. If we start needing to use more training data in the future, we should find a better solution probably.

File: city_metrix/layers/building_classifier/building_classifier.py (hunk 18 above)
Comment (chrowe) [reply city_metrix/layers/building_classifier/building_classifier.py:1]: This is the .py file.

**Comment by chrowe** [comment]
Create models folder and move building_classifier files into it.
